core/types/calls.cc

```diff
@@ -3320,6 +3320,19 @@ class Tuple_concat : public IntrinsicMethod {
     }
 } Tuple_concat;
 
+class Tuple_plus : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        auto *tuple = cast_type<TupleType>(args.thisType);
+        ENFORCE(tuple != nullptr);
+
+        res = dispatchCallProxyType(gs, args.thisType.underlying(gs), args);
+        if (tuple->elems.empty() && args.args.size() == 1) {
+            res.returnType = args.args[0]->type;
+        }
```

Comment:
I think this needs to take into account (with some testcases) what happens if there are errors in the dispatch.  e.g. with the current PR, a test added to `tuple_plus.rb` like:

```ruby
def example4
  xs = []
  T.reveal_type(xs)
  xs += "asdf"
  T.reveal_type(xs)
end
```

gives:

```
test/testdata/infer/tuple_plus.rb:31: Revealed type: [] (0-tuple) https://srb.help/7014
    31 |  T.reveal_type(xs)
          ^^^^^^^^^^^^^^^^^
  Got [] (0-tuple) originating from:
    test/testdata/infer/tuple_plus.rb:30:
    30 |  xs = []
               ^^

test/testdata/infer/tuple_plus.rb:32: Expected T::Enumerable[T.untyped] but found String("asdf") for argument arg0 https://srb.help/7002
    32 |  xs += "asdf"
                ^^^^^^
  Expected T::Enumerable[T.untyped] for argument arg0 of method Array#+:
    https://github.com/sorbet/sorbet/tree/master/rbi/core/array.rbi#L448:
     448 |        arg0: T::Enumerable[Elem],
                  ^^^^
  Got String("asdf") originating from:
    test/testdata/infer/tuple_plus.rb:32:
    32 |  xs += "asdf"
                ^^^^^^

test/testdata/infer/tuple_plus.rb:33: Revealed type: String("asdf") https://srb.help/7014
    33 |  T.reveal_type(xs)
          ^^^^^^^^^^^^^^^^^
  Got String("asdf") originating from:
    test/testdata/infer/tuple_plus.rb:32:
    32 |  xs += "asdf"
          ^^^^^^^^^^^^
```

which seems worse than the current behavior ([sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Adef%20example4%0A%20%20xs%20%3D%20%5B%5D%0A%20%20T.reveal_type%28xs%29%0A%20%20xs%20%2B%3D%20%22asdf%22%0A%20%20T.reveal_type%28xs%29%0Aend%0A)):

```
editor.rb:6: Revealed type: [] (0-tuple) https://srb.help/7014
     6 |  T.reveal_type(xs)
          ^^^^^^^^^^^^^^^^^
  Got [] (0-tuple) originating from:
    editor.rb:5:
     5 |  xs = []
               ^^

editor.rb:7: Expected T::Enumerable[T.untyped] but found String("asdf") for argument arg0 https://srb.help/7002
     7 |  xs += "asdf"
                ^^^^^^
  Expected T::Enumerable[T.untyped] for argument arg0 of method Array#+:
    https://github.com/sorbet/sorbet/tree/master/rbi/core/array.rbi#L448:
     448 |        arg0: T::Enumerable[Elem],
                  ^^^^
  Got String("asdf") originating from:
    editor.rb:7:
     7 |  xs += "asdf"
                ^^^^^^

editor.rb:8: Revealed type: T::Array[T.untyped] https://srb.help/7014
     8 |  T.reveal_type(xs)
          ^^^^^^^^^^^^^^^^^
  Got T::Array[T.untyped] originating from:
    editor.rb:7:
     7 |  xs += "asdf"
          ^^^^^^^^^^^^
Errors: 3
```

Actually, maybe even checking for no errors is not sufficient, because you can do things like:

```
def example5
  xs = []
  T.reveal_type(xs)
  xs += {}
  T.reveal_type(xs)
end
```

which is currently accepted ([sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Adef%20example4%0A%20%20xs%20%3D%20%5B%5D%0A%20%20T.reveal_type%28xs%29%0A%20%20xs%20%2B%3D%20%7B%7D%0A%20%20T.reveal_type%28xs%29%0Aend%0A)), but gives with this PR:

```
test/testdata/infer/tuple_plus.rb:31: Revealed type: [] (0-tuple) https://srb.help/7014
    31 |  T.reveal_type(xs)
          ^^^^^^^^^^^^^^^^^
  Got [] (0-tuple) originating from:
    test/testdata/infer/tuple_plus.rb:30:
    30 |  xs = []
               ^^

test/testdata/infer/tuple_plus.rb:33: Revealed type: {} (shape of T::Hash[T.untyped, T.untyped]) https://srb.help/7014
    33 |  T.reveal_type(xs)
          ^^^^^^^^^^^^^^^^^
  Got {} (shape of T::Hash[T.untyped, T.untyped]) originating from:
    test/testdata/infer/tuple_plus.rb:32:
    32 |  xs += {}
```

which is probably not correct.  But even still, Ruby rejects it:

```
Traceback (most recent call last):
        5: from /home/froydnj/ruby/bin/irb:23:in `<main>'
        4: from /home/froydnj/ruby/bin/irb:23:in `load'
        3: from /home/froydnj/ruby/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        2: from (irb):3
        1: from (irb):3:in `+'
TypeError (no implicit conversion of Hash into Array)
```

Reply:
> But even still, Ruby rejects it

If we want, we can add a special case in the `Tuple_plus` intrinsic to detect not only if the receiver is `Enumerable` but whether it has a `to_ary` method defined.

That plagues what I imagine is quite a few things in the standard library, so I'm inclined to do that some other time.